ansible_base/jwt_consumer/common/auth.py

```diff
@@ -0,0 +1,203 @@
+import logging
+from urllib.parse import urljoin, urlparse
+
+import jwt
+import requests
+from django.contrib.auth import get_user_model
+from rest_framework.authentication import BaseAuthentication
+from rest_framework.exceptions import AuthenticationFailed
+
+from ansible_base.lib.utils.settings import get_setting
+
+logger = logging.getLogger("ansible_base.jwt_consumer.common.auth")
+
+
+# These fields are used to both map the user as well as to validate the JWT token
+default_mapped_user_fields = [
+    "first_name",
+    "last_name",
+    "email",
+    "is_superuser",
+    "is_system_auditor",
+]
+
+
+class JWTCommonAuth:
+    def __init__(self, user_fields=default_mapped_user_fields) -> None:
+        self.mapped_user_fields = user_fields
+
+    def parse_jwt_token(self, request):
+        logger.debug("Starting JWT Authentication")
+        token = request.headers.get("X-DAB-JW-TOKEN", None)
+        if not token:
+            logger.info("X-DAB-JW-TOKEN header not set for JWT authentication")
+            return None, None
+        logger.debug(f"Received JWT auth token: {token}")
+
+        jwt_key_setting = get_setting("ANSIBLE_BASE_JWT_KEY")
+        if not jwt_key_setting:
+            logger.info("Failed to get the setting ANSIBLE_BASE_JWT_KEY")
+            return None, None
+
+        decryption_key = self.get_decryption_key(
+            jwt_key_setting,
+            validate_certs=get_setting("ANSIBLE_BASE_JWT_VALIDATE_CERT", True),
+            timeout=get_setting("ANSIBLE_BASE_JWT_URL_TIMEOUT", 30),
+        )
+        validated_body = self.validate_token(token, decryption_key)
+        user_model = get_user_model()
+        user, created = user_model.objects.update_or_create(
+            username=validated_body["sub"],
+            first_name=validated_body["first_name"],
+            last_name=validated_body["last_name"],
+            email=validated_body["email"],
+            is_superuser=validated_body["is_superuser"],
+        )
+        if created:
+            logger.warn(f"New user {user.username} created from JWT auth")
+        else:
+            logger.info(f"User {user.username} authenticated from JWT auth")
+
+        return user, validated_body
+
+    def log_and_raise(self, details):
+        logger.error(details)
+        raise AuthenticationFailed(details)
+
+    def get_decryption_key_from_url(self, url, timeout, validate_certs):
```

Comment:
I triggered this code path by setting `ANSIBLE_BASE_JWT_KEY = 'https://aap_gw_proxy_1:9080'` BUT Django got pissy because the header `Host: aap_gw_proxy_1` violated some RFC and wouldn't serve my request. I've hit this problem in awx too with container named `tools_awx_1`.

We should bite the bullet and change our container names `s/_/-/g`

```
import socket, requests
ip_add = socket.gethostbyname("aap_gw_proxy_1")
ANSIBLE_BASE_JWT_KEY = requests.get(url=f'https://{ip_add}:9080/api/gateway/v1/jwt_key/', verify=False).text
```
^^ workaround